modules/lang-painless/src/main/java/org/elasticsearch/painless/node/ANode.java

```diff
@@ -44,6 +44,24 @@
  */
 public abstract class ANode {
 
+    protected ANode parent;
+
+    public ANode getParent() {
```

Comment:
Wouldn't keeping bidirectional links make it more difficult to insert nodes between subtrees, eg wrapping an expression in an implicit cast node?